docs/proposals/event-notification.md

```diff
@@ -0,0 +1,143 @@
+
+This is what an `odo push -o json` command invocation would look like:
+```
+odo push -o json
+{ "devFileCommandExecutionBegun": { "commandName" : "build", "timestamp" : "(UTC unix epoch seconds.microseconds)" } }
+{ "devFileActionExecutionBegun" : { "commandName" : "build", "actionName" : "run-build-script", "timestamp" : "(...)" } }
+{ "logText" : { "text:" "one line of text received\\n another line of text received", "timestamp" : "(...)" } }
+{ "devFileActionExecutionComplete" : { "errorCode" : 1, ( same as above )} }
+{ "logText" : { "text": (... ), "timestamp" : "(...)" } } # Log text is interleaved with events 
+{ "devFileCommandExecutionComplete": { "success" : false, (same as above) } }
+
+(Exact details on event name, and JSON format are TBD; feedback welcome!)
+```
+
+These events allow an external odo-consuming tool to determine the build status of an application (build succeeded, build failed, build not running).
+
+Note that unlike other machine-readable outputs used in odo, each individual line is a fully complete and parseable JSON document, allowing events to be streamed and processed by the consuming tool one-at-a-time, rather than waiting for all the events to be received before being parseable (which would be required if the entire console output was one single JSON document, as is the case for other odo machine-readable outputs.)
+
+### Detailed build status via JSON+custom markup
+
+For detailed build status, it is proposed that devfile writers may *optionally* include custom markup in their devfile actions which indicate a detailed build status:
+- If a dev file writer wanted to communicate that the current command/action were compiling the application, they would insert a specific markup string (`#devfile-status#`) at the beginning of a console-outputted line, and then between those two fields would be a JSON object with a single field `buildStatus`:
+  - For example: `#devfile-status# {"buildStatus":"Compiling application"}` would then communicate that the detailed build status should be set to `Compiling application`.
+- Since this line would be output as container stdout, it would be included as a `logText` JSON event, and the consuming tool can look for this markup string and parse the simple JSON to extract the detailed build status.
+- Feedback welcome around exact markup text format.
+
+The build step (running as a bash script, for example, invoked via an action) of a devfile might then look like this:
+```
+#!/bin/bash
+(...)
+echo "#devfile-status# {'buildStatus':'Compiling application'}
+mvn compile
+echo "#devfile-status# {'buildStatus':'Running unit tests'}
+mvn test
+```
+
+This 'detailed build status' markup text is *entirely optional*: if this markup is not present, the odo tool can still determine build succeeded/failed and build running/not-running using the other `odo push -o json` JSON events. 
+
+### App status notification via `odo component status -o json` and `odo log --follow`
```

Comment:
Agreed! Now, what I'm about to say may be a distinction without a useful different, but:  'App status' is definitely an imperfect term for the reasons you described, but I needed a term that also refers to whether the application's OS process is running inside the container or not. In contrast, the term component status is more general; it would refer to whether the pod/container is running, but not (necessarily) to whether the app is running.

These two terms are mostly the same, except in the following cases:
- User's applications fails to start, or dies after starting: In this case the component status is up (eg the pod/container are still running, the odo init volume is mapped in, etc), but the app is down.
- During a devfile build, the app running in the container may be temporarily stopped in order to allow the build to step to deploy new files; in this scenario the app would be down but the component would be up.

In any case, like I said, using app status is problematic, but I'll think a bit more if there's a better term I can use here to refer to both ideas, so that there is no confusion with the odo application/component model terminology 👍 .

Reply:
How about the term "developer application" when we're talking about the application inside the pod/container that a dev cares about? Two words, I know, but unambiguous to a good extent, IMO.

However, broader discussion around the terminology is happening at #3076 & #3190.

Reply:
> However, broader discussion around the terminology is happening at #3076 & #3190.

I would keep the discussion about terminology separate in  #3076. We will need to change and redefine some of the terms used in odo across the odo all commands